src/eins-network-id.c

```diff
@@ -0,0 +1,238 @@
+ * preference default route will appear first)
+ */
+gchar *
+get_ipv4_default_gateway (void)
+{
+  g_autofree gchar *route = NULL;
+
+  if (!g_file_get_contents (PROC_IPV4_ROUTE_FILE, &route, NULL, NULL) || !route)
+    {
+      g_debug ("could not load " PROC_IPV4_ROUTE_FILE);
+      return NULL;
+    }
+
+  g_autoptr(GRegex) route_rx = g_regex_new (IPV4_ROUTE_REGEXP, G_REGEX_MULTILINE, 0, NULL);
+  g_autoptr(GMatchInfo) route_mi = NULL;
+
+  if (!g_regex_match (route_rx, route, 0, &route_mi) || !route_mi)
+    {
+      g_debug ("couldn't match IPv4 default gateway in " PROC_IPV4_ROUTE_FILE);
+      return NULL;
+    }
+
+  const gchar *gateway = g_match_info_fetch_named (route_mi, "Gateway");
+  g_assert (gateway);
+  guint64 gateway_ip = g_ascii_strtoull (gateway, NULL, 16);
+
+  return g_strdup_printf ("%u.%u.%u.%u",
+                          (guint8) gateway_ip,
+                          (guint8) (gateway_ip >> 8),
+                          (guint8) (gateway_ip >> 16),
+                          (guint8) (gateway_ip >> 24));
+}
+
+#define PROC_IPV4_ARP_FILE "/proc/net/arp"
+#define IPV4_ARP_REGEXP "^%s\\s+(\\S+\\s+){2}(?P<HW>[0-9a-f:]{17})"
+
+/*
+ * find the given IPv4 address in the ARP cache to find the HW address.
+ * relies on two assumptions:
+ *  - the first (obviously) is that the IP has been communicated with recently
+ *    enough to be in the cache. for an internet-connected host, this is a
+ *    pretty safe bet for the gateway IP.
+ *  - the second is that if the host has multiple interfaces which are in the
+ *    same IPv4 segment, that these are the same physical network, because the
+ *    search is not constrained by interface which would be the strictly
+ *    correct thing to do. this configuration seems pretty unlikely (two bonded
+ *    load-balanced links with separate gateways?) on a client machine, and for
+ *    our purposes it doesn't matter if we always report one gateway, or report
+ *    both over time, so it's not worth getting excited about.
+ */
+gchar *
+get_ipv4_hwaddr (const gchar *address)
+{
+  g_autofree gchar *arp = NULL;
+
+  if (!g_file_get_contents (PROC_IPV4_ARP_FILE, &arp, NULL, NULL) || !arp)
+    {
+      g_debug ("could not load " PROC_IPV4_ARP_FILE);
+      return NULL;
+    }
+
+  g_autofree gchar *arp_rx_raw = g_strdup_printf (IPV4_ARP_REGEXP, address);
+  g_autoptr(GRegex) arp_rx = g_regex_new (arp_rx_raw, G_REGEX_MULTILINE, 0, NULL);
+  g_autoptr(GMatchInfo) arp_mi = NULL;
+
+  if (!g_regex_match (arp_rx, arp, 0, &arp_mi) || !arp_mi)
+    {
+      g_debug ("couldn't find HW address in " PROC_IPV4_ARP_FILE);
+      return NULL;
+    }
+
+  const gchar *hw = g_match_info_fetch_named (arp_mi, "HW");
+  g_assert (hw);
+  return g_ascii_strdown (hw, -1);
+}
+
+#define PROC_IPV6_ROUTE_FILE "/proc/net/ipv6_route"
+#define IPV6_ROUTE_REGEXP "^(0{32}\\s00\\s){2}(?P<Gateway>[0-9a-f]{32})"
+
+/*
+ * greps the /proc/net/ipv6_route file for the first IPv6 default route. same
```

Comment:
I checked, as per the IPv4 route file it's a walk of the tree which is ordered by of prefix and metric which should be stable enough - ethernet always beats wifi, for example. The first "hit" that the kernel makes when routing a packet via the gateway will be our first hit too.

Although, in checking this I did find a bug in picking up routes for 0000::/0 which were not via gateways, so I extended the regexp to check the flags show the route is RTF_UP (0x1) and involves a RTF_GATEWAY (0x2). RTF_HOST (0x4) should be mutually exclusive with the default youre, and RTF_REINSTATE (0x8) is unused in Linux so I don't think we need to worry about being more clever than looking for a trailing 3.